### PR TITLE
[checkpoint] Run syncing in active process for fullnodes

### PR DESCRIPTION
**Pull request by williampsmith**
This adopts a similar pattern to the handling of the `sui-node` gossip process handle (which, for full nodes is a process to sync state from a quorum of validators, and for validators, is an active gossip process for txes, etc.), but for the checkpoint process.

Before: 
* Full nodes do a full checkpoint sync once before spawning an active process for syncing state with validators
* Validator nodes spawn an active process for syncing and driving checkpoints

After:
* Both full nodes and validators spawn an active checkpoint process. 
* For validators, this is the same checkpoint driving process as before
* For full nodes, this is an active checkpoint syncing process that runs periodically

Other notes on this change:
* In order to facilitate testing checkpoints on a simtest swarm, it was necessary to make `CheckpointProcessControl` configurable so as to avoid waiting 2+ minutes for the pause between checkpoint times. `CheckpointProcessControl` is now configured by setting a member of the `Node`, and for simtests that call `start_a_fullnode*`, is configured to wait on the order of seconds between checkpoints.
* As such, we can also now configure CheckpointProcessControl from the `NetworkConfig` yaml

### CODE REVIEW

File: crates/sui-core/src/authority_active/checkpoint_driver/mod.rs [hunk 1]
```diff
@@ -370,6 +231,192 @@ where
...
+    let state_checkpoints = &active_authority.state.checkpoints;
+
+    if let Some(checkpoint) = highest_checkpoint {
+        debug!(
+            "Highest Checkpoint Certificate from the network: {}",
+            checkpoint
+        );
+        // Check if there are more historic checkpoints to catch up with
+        let next_checkpoint = state_checkpoints.lock().next_checkpoint();
+        if next_checkpoint < checkpoint.summary.sequence_number {
```
Comment (lxfind) [inline crates/sui-core/src/authority_active/checkpoint_driver/mod.rs:292]: This will make fullnode always one checkpoint behind, right?

**Review by lxfind — COMMENTED** [review]
Ideally the sync process should always run regardless of whether it's a fullnode or validator. The validator process then does not need to worry about sync.
This will be less efficient in the short run, but it helps setup the code in the right structure, as that's what it would be down the road.
Also, we may want to separate the state to two parts: one that sync cares about, and one that the driver cares about. Although a lot of this will likely change once @bmwill 's change is in, so I am not very sure what is the best way to proceed.

File: crates/sui-core/src/authority_active/checkpoint_driver/mod.rs (hunk 1 above)
Comment (williampsmith) [reply crates/sui-core/src/authority_active/checkpoint_driver/mod.rs:292]: @lxfind yeah I initially thought this might be the case, but was unsure since this is the same bound that the fullnode runs for its one-time startup sync... I assumed that the reason it is +1 for a validator is because a validator would be participating in the current checkpoint, whereas the fullnode is just syncing already finished checkpoints. Am I misinterpreting these bounds?

Comment (lxfind) [reply crates/sui-core/src/authority_active/checkpoint_driver/mod.rs:292]: If you read the comments for `update_latest_checkpoint`, it's because for validators, the latest checkpoint cert could either be from some validator, or formed on-the-fly.
For fullnode, however, we shouldn't be doing the forming checkpoint cert on the fly. It should strictly try to find the latest cert that some validator in the network has.

File: crates/sui/tests/full_node_tests.rs [hunk 2]
```diff
@@ -86,6 +86,27 @@ async fn test_full_node_follows_txes() -> Result<(), anyhow::Error> {
     Ok(())
 }
 
+#[sim_test]
+async fn test_full_node_syncs_checkpoints() -> Result<(), anyhow::Error> {
+    let test_cluster = init_cluster_builder_env_aware().build().await?;
+    let sui_node: SuiNode = start_a_fullnode(&test_cluster.swarm, false).await?;
+    let node = if cfg!(msim) {
+        &sui_node
```
Comment (mystenmark) [inline crates/sui/tests/full_node_tests.rs:94]: Why was this necessary? the `SuiNode` api should not change in the simulator.

File: crates/sui-core/src/authority_active/checkpoint_driver/mod.rs [hunk 3]
```diff
@@ -370,6 +231,192 @@ where
...
+                        return;
+                    } else if active_authority.state.is_fullnode()
+                        && state_checkpoints.lock().is_ready_to_finish_epoch_change()
+                    {
+                        while let Err(err) = active_authority.finish_epoch_change().await {
+                            error!(?next_cp_seq, "Failed to finish epoch change: {:?}", err);
+                            tokio::time::sleep(timing.epoch_change_retry_delay).await;
+                        }
+                    }
+                    tokio::time::sleep(timing.long_pause_between_checkpoints).await;
```
Comment (mystenmark) [inline crates/sui-core/src/authority_active/checkpoint_driver/mod.rs:360]: if i'm reading this right, fullnode will take a long pause after syncing every checkpoint. but we want fullnode to sync as fast as it can. But, when this is part of the checkpoint creation step (on validator), we do indeed want to pause after a new checkpoint is formed.

Comment (williampsmith) [reply crates/sui-core/src/authority_active/checkpoint_driver/mod.rs:360]: Makes sense for new node startup syncing, but for incremental syncs of new checkpoints, won't removing this cause a ton of busy looping once there's a new checkpoint?

File: crates/sui/tests/full_node_tests.rs (hunk 2 above)
Comment (williampsmith) [reply crates/sui/tests/full_node_tests.rs:94]: What exactly are you referring to?

**Comment by williampsmith** [comment]
Abandoning in light of checkpoint v2
